### PR TITLE
[rllib] Generalizing A3C Sampling Classes

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?
The A3C implementation has nice support around two things that other algorithms currently do not have - LSTM/RNN support and Partial Rollouts. This PR will retool the A3C classes so that they can be easily used by other algorithms.

 - Adds support for Observation Filters in AsyncSampler
 - Allows for varying output values in `compute_action` (Actor Critics require `value`, RNNs require `features`, option critics require `options`, PPO requires `logprobs`) by adapting `PartialRollouts` and `process_rollout` to support this, and adding `other_fields` to the models
    - @ericl can you take a closer look at this part? I'm not too certain about the design.
 - [x] Add support for Reward filters (what are examples of reward filters..?)
 - [x] Better documentation
 - [x] Generic Samplers (synchronous)
 
### Related but out of scope for this PR:
 - Removing BatchedEnvs, Adding RNN support for PPO
 - Generalizing advantage estimation post-processing of rollout data

### CODE REVIEW

File: python/ray/rllib/a3c/common.py [hunk 1]
```diff
@@ -5,31 +5,59 @@
...
+        batch_adv[i] = reward_filter(batch_adv[i])
 
-    features = rollout.features[0]
-    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.terminal,
+    features = rollout.data["features"][0]
+    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.is_terminal(),
                  features)
 
 
+def get_filter(filter_config, shape):
```
Comment (ericl) [inline python/ray/rllib/a3c/common.py]: I think it's time we moved PPO's filters to a common place (both DQN+A3C use it now). How about `rllib/util`?

**Review by ericl — COMMENTED** [review]
> Allows for varying output values in compute_action (Actor Critics require value, RNNs require features, option critics require options, PPO requires logprobs) by adapting PartialRollouts and process_rollout to support this, and adding other_fields to the models

The best way to see if the abstractions make sense would be to try porting some other algorithms to use it (e.g. PPO or DQN) and see if it simplifies the code.

> Adds support for Observation Filters in AsyncSampler

If the proposal is to standardize on one observation filter API, can we move the filters to a common directory?

File: python/ray/rllib/utils/filter.py [hunk 2]
```diff
@@ -128,7 +172,7 @@ def __call__(self, x, update=True):
             else:
                 # The unvectorized case.
                 self.rs.push(x)
-                self.buffer.push(x[i])
+                self.buffer.push(x)
```
Comment (richardliaw) [inline python/ray/rllib/utils/filter.py]: This was a bug that somehow never got caught

**Comment by richardliaw** [comment]
> The best way to see if the abstractions make sense would be to try porting some other algorithms to use it (e.g. PPO or DQN) and see if it simplifies the code.

This should be done in PPO with #1253.

> ...  can we move the filters to a common directory?

Done.

File: python/ray/rllib/a3c/runner.py [hunk 3]
```diff
@@ -2,82 +2,75 @@
...
+            config["observation_filter"], env.observation_space.shape)
+        self.rew_filter = get_filter(config["reward_filter"], ())
+
+        # TODO(rliaw): Convert this to a cataloged object.
+        self.sampler = AsyncSampler(env, self.policy, config["batch_size"],
+                                    obs_filter)
         self.logdir = logdir
-        self.start()
+        if self.sampler.async:
+            self.sampler.start_runner()
```
Comment (richardliaw) [inline python/ray/rllib/a3c/runner.py]: This can probably be pushed into the __init__ of sampler

File: python/ray/rllib/utils/filter.py [hunk 4]
```diff
@@ -5,19 +5,39 @@
 import numpy as np
 
 
-class NoFilter(object):
-    def __init__(self):
+class BaseFilter(object):
+
+    def update(self, other, *args, **kwargs):
+        """Given other filter, update self with "new state" from
+        other filter. This may be in form of a buffer.
```
Comment (ericl) [inline python/ray/rllib/utils/filter.py]: https://www.python.org/dev/peps/pep-0257/#id17

File: python/ray/rllib/utils/filter.py [hunk 5]
```diff
@@ -5,19 +5,39 @@
 import numpy as np
 
 
-class NoFilter(object):
-    def __init__(self):
+class BaseFilter(object):
+
```
Comment (ericl) [inline python/ray/rllib/utils/filter.py]: Describe what filters do?

File: python/ray/rllib/utils/filter.py [hunk 6]
```diff
@@ -5,19 +5,39 @@
...
+        """Given other filter, update self with "new state" from
+        other filter. This may be in form of a buffer.
+        """
+        raise NotImplementedError
+
+    def copy(self):
+        """Creates a new object with same state as self. Returns copy."""
+        raise NotImplementedError
+
+    def sync(self, other):
```
Comment (ericl) [inline python/ray/rllib/utils/filter.py]: why not set `filter = other.copy()`?

File: python/ray/rllib/utils/sampler.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        completed = []
+        while True:
+            try:
+                completed.append(self.metrics_queue.get_nowait())
+            except queue.Empty:
+                break
+        return completed
+
+
+def env_runner(env, policy, num_local_steps, obs_filter):
```
Comment (ericl) [inline python/ray/rllib/utils/sampler.py]: _env_runner

File: python/ray/rllib/utils/sampler.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+        with lock:
+            return func(*args, **kwargs)
+    return wrapper
+
+
+class PartialRollout(object):
+    """A piece of a complete rollout.
+
+    We run our agent, and process its experience once it has processed enough
+    steps.
```
Comment (ericl) [inline python/ray/rllib/utils/sampler.py]: What returns a partial rollout?

File: python/ray/rllib/utils/sampler.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        Returns:
+            terminal (bool): if rollout has terminated."""
+        return self.data["terminal"][-1]
+
+
+CompletedRollout = namedtuple(
+    "CompletedRollout", ["episode_length", "episode_reward"])
+
+
+class SyncSampler(object):
```
Comment (ericl) [inline python/ray/rllib/utils/sampler.py]: Does this and AsyncSampler implement any common interface?

File: python/ray/rllib/utils/filter.py (hunk 6 above)
Comment (richardliaw) [reply python/ray/rllib/utils/filter.py]: For async reasons... `lock_wrap` takes the `MeanStdFilter` and passes it to the other thread. Changing the reference won't update the object in `lock_wrap`. Suggestions welcome here though.

File: python/ray/rllib/utils/sampler.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/rllib/utils/sampler.py]: Yeah, should we make a class for that? I feel like there won't be any other Sampler classes though.

File: python/ray/rllib/utils/sampler.py (hunk 8 above)
Comment (richardliaw) [reply python/ray/rllib/utils/sampler.py]: `_env_runner`

**Review by ericl — APPROVED** [review]

File: python/ray/rllib/a3c/common.py [hunk 10]
```diff
@@ -11,27 +11,41 @@ def discount(x, gamma):
...
     # This formula for the advantage comes "Generalized Advantage Estimation":
     # https://arxiv.org/abs/1506.02438
     batch_adv = discount(delta_t, gamma * lambda_)
+    for i in range(batch_adv.shape[0]):
+        batch_adv[i] = reward_filter(batch_adv[i])
 
-    features = rollout.features[0]
-    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.terminal,
+    features = rollout.data["features"][0]
+    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.is_terminal(),
```
Comment (ericl) [inline python/ray/rllib/a3c/common.py]: this is still wip, but I'd like to replace all sample batches with a standard class https://github.com/ericl/ray/blob/evaluator-multi-gpu/python/ray/rllib/optimizers/sample_batch.py to better support pluggable optimizers.

Comment (richardliaw) [reply python/ray/rllib/a3c/common.py]: https://github.com/openai/baselines/blob/master/baselines/common/dataset.py

Check this out - this is a lot cleaner than what's going on right now imo

**Pull request merged**